### ISSUE TITLE
Form fields are only considered touched when they were blurred (at least) once

### ISSUE THREAD

**Issue by lennartbuit** [reakit/reakit#640]
### Current behavior

There is a difference in behaviour of validations between fields that have been blurred, and fields that have not. Validations do run on fields that haven't been blurred yet, but the field is only marked touched when the field loses focus, leading errors to only show up for fields that were blurred at least once.

### Steps to reproduce the bug

(Not providing codesandbox, as the Reakit documentation shows the behaviour) Take the usage notes for `Form`, here: https://reakit.io/docs/form/#usage

1. Focus the `name` field
2. Add and remove a name => no error messages
3. Blur the `name` field
4. Focus the `name` field
2. Add and remove a name => there are now error messages

### Expected behavior

Behaviour should not be different between fields that were and were not blurred once. 

### Possible solutions

One possibility that is applicable to my use case would be to mark fields touched when they were changed, instead of when they were blurred. That would (I believe) make behaviour of fields consistent between being blurred at least once / not being blurred.

Alternatively, touchedness can be tied to when validations run. If they run on blur, then touched can also be set onBlur, whereas if they run on change, then touched can (also) be set onChange.

**Comment by lennartbuit** [comment]
FWIW: [this is where fields are marked touched](https://github.com/reakit/reakit/blob/0c51c4f3dc1ec9a83b9d915039ad8167f4e4217b/packages/reakit/src/Form/FormState.ts#L250), which indeed only runs on blur.

Also; if we reach some consensus on how this should work, I would also like to offer to implement a solution myself. Reakit is of great help to me, so I'd like to contribute where I can!

**Comment by diegohaz** [comment]
Hi @lennartbuit! Thanks for opening the issue.

This is working as intended, and the way it works is the same as in other libraries, like Formik.

However, we can think on an option to mark fields as touched when they're updated, for example. Or something else. If you describe your use case, it may help us find a better API.

**Comment by lennartbuit** [comment]
Hey there, thank you for your (swift!) response. The reason I am asking is because I abstracted the form state away (like suggested in the docs) and have sub components that use a shared context. One of these components is a submit button, I use the form state to determine whether the form can be submitted and thus enabling the submit button. However, as far as I can tell, validations only run when a change was made (this is great! don't want to show errors when people didn't do anything yet), and a form starts out valid. Therefore I am using the touchedness to determine whether it is valid because it wasn't touched yet.

The problem I am now ending up with is that in a single field form, changing a field does run the validations, but does not update the touchedness. This leads me to not be able to determine whether to enable the submit button, no fields are officially touched, and the validations succeeded, so I cannot determine a difference with the initial state.

Since you mentioned Formik, I don't have any experience with it, but I took a look at how it differs. [They offer a code sandbox as well](https://codesandbox.io/s/zKrK5YLDZ?file=/index.js:1410-1422), and this example - incidentally - uses `dirty` to enable the reset button. The difference, however, is that their dirty checking does run on change, and compares initial state with the current.

Once more, thank you (all) for this great library!

**Comment by diegohaz** [comment]
So, just to avoid confusion, there are two concepts here: running the validation and actually showing the errors.

By default, validations run on change and on blur, and you can configure it by passing `validateOnChange` and `validateOnBlur` to `useFormState()`. They're `true` by default:
https://github.com/reakit/reakit/blob/0c51c4f3dc1ec9a83b9d915039ad8167f4e4217b/packages/reakit/src/Form/FormState.ts#L285-L286

The error is shown only after the field has been touched (on blur), so you don't show errors while the user is still typing (imagine that a field has minimum length of 3, and as soon as the user types the first letter you show them an error message 😬):
https://github.com/reakit/reakit/blob/0c51c4f3dc1ec9a83b9d915039ad8167f4e4217b/packages/reakit/src/Form/__utils/shouldShowError.ts#L8

So, validations may run and set `form.valid` and `form.errors` even when the errors aren't being displayed to the user. In order to disable the submit button when the form is invalid, you could do this:
```jsx
const form = useFormState();
<FormSubmitButton {...form} disabled={!form.valid} />
```

Validations don't run on mount, so the button would be enabled until some form value is changed. We can add a new `validateOnMount` option similar to [the Formik one](https://jaredpalmer.com/formik/docs/api/formik#validateonmount-boolean). But this can also be achieved in the user land:
```js
const form = useFormState();
// Guarantees that we'll only use the first version of form.validate(),
// and our effect will not run on updates
const [initialValidate] = React.useState(() => form.validate);
React.useEffect(() => {
  initialValidate().catch(() => {});
}, [initialValidate]);
```

Here's a CodeSandbox with `dirty` and `validateOnMount`: https://codesandbox.io/s/reakit-form-validateonmount-and-dirty-state-td8ph?file=/index.js

If they're useful, I'm totally open to adding them to the library. :)

**Comment by lennartbuit** [comment]
Hey, thank you for your suggestion, works perfectly :). Here are some things I noticed while integrating them. I hope that they help in determining what to include in Reakit!

1. I did not end up needing dirty tracking

The reason being, a 'create' like form is (usually) not valid from the get-go. For 'update' forms, I thought it would be confusing to users that a perfectly valid form could not be submitted. 

In other words, I think the value of preventing a superfluous submit is less than the confusion it creates when a perfectly valid form could not be submitted because it was not touched.

2. `validateOnMount` runs after first render, which can cause this submit button to flash.

`valid` is true by default, so prior to the validations running on mount, this submit button is enabled. Basically, in 'create' forms, I expect the form to not be valid initially, whereas in 'update' forms I do. I am currently remedying this by keeping track of whether validations ran, and if they didn't return a prop with the 'default' validness. 

Looking at Formik, I see they have deprecated this default validness. So maybe thats something that has a prettier solution.

-----

So, for my usecase, I'd say that `validateOnMount` would be helpful, but dirty tracking can be omitted!
